### PR TITLE
Add checking of the iRODS path type

### PR DESCRIPTION
**Pull request by kjsanger**
Raise an error if a DataObject is created from a collection path
Raise an error if a Collection is created from a data object path
Add a decorator to apply the check to appropriate methods
Add a constructor parameter to enable the old behaviour of not checking

### CODE REVIEW

File: src/partisan/irods.py [hunk 1]
```diff
@@ -1046,11 +1045,38 @@ def __str__(self):
...
-    def __init__(self, path: Union[PurePath, str], pool=default_pool):
+    def __init__(self, path: Union[PurePath, str], check_type=False, pool=default_pool):
+        """
+        RodsItem constructor.
+        Args:
+            path: A remote path.
+            check_type: Check the remote path type if True, defaults to False.
+            pool: A baton client pool. Optional.
+        """
         self.path = PurePath(path)
```
Comment (marcomoscasgr) [inline src/partisan/irods.py:1077]: Why is the abstract class PurePath used? Is that related to some later features compatible to other operative systems?

**Review by marcomoscasgr — COMMENTED** [review]
Clarifications on PurePath and function inheritance for DataObject and Collection classes

File: src/partisan/irods.py (hunk 1 above)
Comment (kjsanger) [reply src/partisan/irods.py:1077]: `PurePath` is not abstract in the object-oriented programming sense. "Pure path objects provide path-handling operations which don’t actually access a filesystem." (see https://docs.python.org/3/library/pathlib.html). This makes them perfect for representing iRODS paths, which are virtual. A data object has a path stored in the iRODS database which is not the same as the path(s) to the files(s) holding the data.

File: src/partisan/irods.py [hunk 2]
```diff
@@ -1389,6 +1445,21 @@ def query_metadata(
...
+
+        if not self._exists(timeout=timeout, tries=tries):
+            return
+
+        item = self._list(timeout=timeout, tries=tries).pop()
+        if Baton.OBJ not in item.keys():
+            raise BatonError(
+                f"Invalid iRODS path for a data object; the {Baton.OBJ} "
+                f"key is not in {item}, indicating a collection"
+            )
```
[the remark this answers is not in the document]
Comment (kjsanger) [reply src/partisan/irods.py:1460]: The checks necessary for `DataObject` and `Collection` are different; one checks that a key _is_ present, the other that it's _not_ present in the JSON returned from the client. That's why I've made the method polymorphic.

**Review by marcomoscasgr — APPROVED** [review]

**Pull request merged**